jax/tree_util.py

```diff
@@ -172,28 +171,66 @@ def __ne__(self, other):
 
 
 class PyLeaf(object):
+  include_subclasses = True
   def __repr__(self):
     return '*'
 
 leaf = PyLeaf()
 
 def dict_to_iterable(xs):
-  keys = tuple(sorted(xs.keys()))
-  return tuple(map(xs.get, keys)), keys
+  keys, values = zip(*sorted(xs.items()))
+  return values, keys
 
 class NodeType(object):
-  def __init__(self, name, to_iterable, from_iterable):
+  def __init__(self, name, to_iterable, from_iterable, include_subclasses):
     self.name = name
     self.to_iterable = to_iterable
     self.from_iterable = from_iterable
+    self.include_subclasses = include_subclasses
 
 node_types = {}
 
-def register_pytree_node(py_type, to_iterable, from_iterable):
+def get_node_type(py_type):
+  for cls in py_type.__mro__:
+    node_type = node_types.get(cls)
+    if node_type is not None and (
+        cls is py_type or node_type.include_subclasses):
+      return node_type
+  return leaf
+
+def register_pytree_node(py_type, to_iterable, from_iterable,
+                         include_subclasses=True):
+  """Registers a Python type as a container-like data structure.
+
+  Many JAX functions work equally well with individual arrays and
+  nested trees of tuples, lists, dicts and other containers. This
+  function registers a new container data type with the tree-handling
+  machinery that enables that functionality. If multiple handlers
+  are registered that apply to the same type, the most specific one
+  is used.
+
+  Args:
+    py_type: A type object, such as a user-defined class.
+    to_iterable: A function that takes an instance of that type and
+      returns an iterable containing the instance's child objects
+      and an object holding other state needed to rebuild the
+      instance.
+    from_iterable: A function that takes a state object returned by
+      the to_iterable function, plus an iterable, and returns an
+      instance of the py_type type or a subclass.
+    include_subclasses: Whether the handlers being registered should
+      also apply to subclasses of the provided type.
+  """
   assert py_type not in node_types
-  node_types[py_type] = NodeType(str(py_type), to_iterable, from_iterable)
+  node_types[py_type] = NodeType(
+    str(py_type), to_iterable, from_iterable, include_subclasses)
 
-register_pytree_node(tuple, lambda xs: (xs, None), lambda _, xs: tuple(xs))
+def register_pytree_leaf(py_type):
```

Comment:
Is `register_pytree_leaf` just optional?

It seems like it might be redundant: if we just have node types registered, then everything that isn't registered is a leaf. If we register both leaves and nodes, then in principle now unregistered things are a third thing (?), since they're neither a leaf nor a node.

I might be misunderstanding though. Can you explain the logic?

If indeed registering leaves is optional and does potentially break the current "registered = node, unregistered = leaf" dichotomy, then maybe we can remove it. What do you think?